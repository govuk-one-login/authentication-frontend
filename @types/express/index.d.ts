import type pino from "pino";
import type { Taxonomy } from "../utils/taxonomy.ts";

declare global {
  namespace Express {
    export interface Request {
      i18n?: {
        language?: string;
      };
      t: TFunction;
      csrfToken?: () => string;
      log: pino.Logger;
      generatedSessionId?: string;
    }

    export interface Locals {
      // Session IDs
      sessionId?: string;
      clientSessionId?: string;
      persistentSessionId?: string;
      clientId?: string;

      // Security
      scriptNonce?: string;
      csrfToken?: string;

      // Analytics
      ga4ContainerId?: string;
      isGa4Enabled?: string;
      analyticsCookieDomain?: string;
      analyticsProperties?: {
        httpStatusCode: number;
        contentId: string;
        taxonomy: Taxonomy;
      };

      // Language
      languageToggleEnabled?: boolean;
      language?: string;
      htmlLang?: string;

      // App
      strategicAppChannel?: boolean;
      webChannel?: boolean;
      genericAppChannel?: boolean;
      isApp?: boolean;

      // Feature flags
      supportPasskeyUsage?: boolean;
      supportPasskeyRegistration?: boolean;

      // Misc
      currentUrl?: URL;
      showTestBanner?: boolean;
      accountManagementUrl?: string;
      contactUsLinkUrl?: string;
    }
  }
}
