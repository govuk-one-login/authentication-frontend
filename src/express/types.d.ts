import 'express-serve-static-core';

declare global {
  namespace Express {
    interface Locals {
      sessionId: string;
      strategicAppChannel: boolean;
      webChannel: boolean;
      genericAppChannel: boolean;
      isApp: boolean
    }
    interface Response {
      locals: Locals;
    }
  }
}
