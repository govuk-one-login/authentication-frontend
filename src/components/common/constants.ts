import { PATH_NAMES } from "../../app.constants";
import { getNextState } from "./state-machine/state-machine";

export const SECURITY_CODE_ERROR = "actionType";

export enum SecurityCodeErrorType {
  MfaMaxCodesSent = "mfaMaxCodesSent",
  MfaBlocked = "mfaBlocked",
  MfaMaxRetries = "mfaMaxRetries",
  OtpMaxCodesSent = "otpMaxCodesSent",
  OtpBlocked = "otpBlocked",
  OtpMaxRetries = "otpMaxRetries",
  EmailMaxCodesSent = "emailMaxCodesSent",
  EmailBlocked = "emailBlocked",
  EmailMaxRetries = "emailMaxRetries",
  AuthAppMfaNoCodeValidator = "authAppMfaNoCodeValidator",
  AuthAppMfaMaxRetries = "authAppMfaMaxRetries",
  ChangeSecurityCodesEmailMaxCodesSent = "changeSecurityCodesEmailMaxCodesSent",
  ChangeSecurityCodesEmailBlocked = "changeSecurityCodesEmailBlocked",
  ChangeSecurityCodesEmailMaxRetries = "changeSecurityCodesEmailMaxRetries",
  InvalidPasswordResetCodeMaxRetries = "invalidPasswordResetCodeMaxRetries",
}

export const ERROR_CODES = {
  SESSION_ID_MISSING_OR_INVALID: 1000,
  RESET_PASSWORD_LINK_MAX_RETRIES_REACHED: 1022,
  RESET_PASSWORD_LINK_BLOCKED: 1023,
  NEW_PASSWORD_SAME_AS_EXISTING: 1024,
  ENTERED_INVALID_MFA_MAX_TIMES: 1027,
  VERIFY_EMAIL_CODE_REQUEST_BLOCKED: 1029,
  VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED: 1030,
  VERIFY_PHONE_NUMBER_MAX_CODES_SENT: 1032,
  VERIFY_EMAIL_MAX_CODES_SENT: 1031,
  ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES: 1033,
  ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES: 1034,
  MFA_SMS_MAX_CODES_SENT: 1025,
  MFA_CODE_REQUESTS_BLOCKED: 1026,
  INVALID_MFA_CODE: 1035,
  INVALID_VERIFY_EMAIL_CODE: 1036,
  INVALID_VERIFY_PHONE_NUMBER_CODE: 1037,
  ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES: 1039,
  INVALID_PASSWORD_MAX_ATTEMPTS_REACHED: 1028,
  RESET_PASSWORD_INVALID_CODE: 1021,
  PASSWORD_IS_COMMON: 1040,
  AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED: 1042,
  AUTH_APP_INVALID_CODE: 1043,
  ACCOUNT_LOCKED: 1045,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT: 1046,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED: 1047,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE: 1048,
};

export const ERROR_CODE_MAPPING: { [p: string]: string } = {
  [ERROR_CODES.ACCOUNT_LOCKED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.INVALID_PASSWORD_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.MFA_SMS_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxCodesSent
  ),
  [ERROR_CODES.MFA_CODE_REQUESTS_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_MFA_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpMaxCodesSent
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.OtpMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxRetries
  ),
  [ERROR_CODES.ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.InvalidPasswordResetCodeMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxCodesSent
  ),
  [ERROR_CODES.AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.AuthAppMfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxCodesSent
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxRetries
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_WAIT"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailBlocked
    ),
};

export function getErrorPathByCode(errorCode: number): string | undefined {
  const nextPath = ERROR_CODE_MAPPING[errorCode.toString()];

  if (!nextPath) {
    return undefined;
  }

  return nextPath;
}

export function pathWithQueryParam(
  path: string,
  queryParam?: string,
  value?: string | SecurityCodeErrorType
): string {
  if (queryParam && value) {
    const queryParams = new URLSearchParams({
      [queryParam]: value,
    }).toString();

    return path + "?" + queryParams;
  }

  return path;
}

export function getNextPathAndUpdateJourney(
  req: any,
  path: string,
  event: string,
  ctx?: any,
  sessionId?: string
): string {
  const nextState = getNextState(path, event, ctx);

  req.session.user.journey = {
    nextPath: nextState.value,
    optionalPaths:
      Object.keys(nextState.meta).length > 0
        ? nextState.meta["AUTH." + nextState.value].optionalPaths
        : [],
  };

  req.log.info(
    `User journey transitioned from ${req.path} to ${nextState.value} with session id ${sessionId}`
  );

  if (!nextState) {
    throw Error(
      `Invalid user journey. No transition found from ${path} with event ${event}`
    );
  }

  return nextState.value;
}

export const JOURNEY_TYPE = {
  SIGN_IN: "sign-in",
  CREATE_ACCOUNT: "create-account",
};
