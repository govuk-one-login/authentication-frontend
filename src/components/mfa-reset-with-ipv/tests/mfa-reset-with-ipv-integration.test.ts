import { describe } from "mocha";
import { request, sinon } from "../../../../test/utils/test-utils.js";
import decache from "decache";
import { API_ENDPOINTS, CHANNEL, PATH_NAMES } from "../../../app.constants.js";
import nock from "nock";
import { NextFunction, Request, Response } from "express";
import supertest from "supertest";

const IPV_DUMMY_URL =
  "https://test-idp-url.com/callback?code=123-4ddkk0sdkkd-ad";

describe("Mfa reset with ipv", () => {
  const REQUEST_PATH = PATH_NAMES.MFA_RESET_WITH_IPV;

  describe("Mfa reset with ipv get", () => {
    beforeEach(() => {
      nock.cleanAll();
    });

    after(() => {
      sinon.restore();
      delete process.env.SUPPORT_MFA_RESET_WITH_IPV;
      delete process.env.ROUTE_USERS_TO_NEW_IPV_JOURNEY;
    });

    it("should redirect to the mfa reset ipv path when coming from sms entry", async () => {
      const previousPath = PATH_NAMES.ENTER_MFA;
      const app = await setupAppWithSessionMiddleware(
        previousPath,
        REQUEST_PATH,
        true,
        CHANNEL.WEB
      );
      allowCallToMfaResetAuthorizeEndpointReturningAuthorizeUrl(IPV_DUMMY_URL);

      await request(
        app,
        (test) =>
          test.get(REQUEST_PATH).expect(302).expect("Location", IPV_DUMMY_URL),
        { expectAnalyticsPropertiesMatchSnapshot: false }
      );
    });

    it("should redirect to the mfa reset ipv path when coming from auth app entry", async () => {
      const previousPath = PATH_NAMES.ENTER_AUTHENTICATOR_APP_CODE;
      const app = await setupAppWithSessionMiddleware(
        previousPath,
        REQUEST_PATH,
        true,
        CHANNEL.WEB
      );
      allowCallToMfaResetAuthorizeEndpointReturningAuthorizeUrl(IPV_DUMMY_URL);

      await request(
        app,
        (test) =>
          test.get(REQUEST_PATH).expect(302).expect("Location", IPV_DUMMY_URL),
        { expectAnalyticsPropertiesMatchSnapshot: false }
      );
    });

    it("should redirect to the new guidance page when using the strategic app", async () => {
      const previousPath = PATH_NAMES.ENTER_AUTHENTICATOR_APP_CODE;
      const app = await setupAppWithSessionMiddleware(
        previousPath,
        REQUEST_PATH,
        true,
        CHANNEL.STRATEGIC_APP
      );

      const expectedGuidancePage = PATH_NAMES.OPEN_IN_WEB_BROWSER;

      const agent = supertest.agent(app);

      await agent
        .get(REQUEST_PATH)
        .set("Cookie", "channel=strategic_app")
        .expect(302)
        .expect("Location", expectedGuidancePage);

      await agent.get(expectedGuidancePage).expect(200);
    });

    it("should return a 500 if account recovery is not permitted", async () => {
      const previousPath = PATH_NAMES.ENTER_MFA;
      const app = await setupAppWithSessionMiddleware(
        previousPath,
        REQUEST_PATH,
        false,
        CHANNEL.WEB
      );

      await request(app, (test) => test.get(REQUEST_PATH).expect(500), {
        expectAnalyticsPropertiesMatchSnapshot: false,
      });
    });
  });

  describe("Open One login in browser get", () => {
    const REQUEST_PATH = PATH_NAMES.OPEN_IN_WEB_BROWSER;
    before(() => (process.env.SUPPORT_MFA_RESET_WITH_IPV = "1"));

    after(() => {
      sinon.restore();
      delete process.env.SUPPORT_MFA_RESET_WITH_IPV;
    });

    it("should not render the page when coming from an arbitrary page", async () => {
      const previousPath = PATH_NAMES.AUTHORIZE;

      const app = await setupAppWithSessionMiddleware(
        previousPath,
        REQUEST_PATH,
        true,
        CHANNEL.STRATEGIC_APP
      );

      await request(
        app,
        (test) =>
          test
            .get(PATH_NAMES.OPEN_IN_WEB_BROWSER)
            .expect(302)
            .expect("Location", previousPath),
        { expectAnalyticsPropertiesMatchSnapshot: false }
      );
    });
  });

  function allowCallToMfaResetAuthorizeEndpointReturningAuthorizeUrl(
    url: string
  ) {
    const baseApi = process.env.FRONTEND_API_BASE_URL;

    nock(baseApi).post(API_ENDPOINTS.MFA_RESET_AUTHORIZE).once().reply(200, {
      authorize_url: url,
      code: 200,
      success: true,
    });
  }

  async function setupAppWithSessionMiddleware(
    currentNextPath: string,
    firstRequestPath: string,
    isAccountRecoveryPermitted: boolean,
    channel: string
  ) {
    decache("../../../app");
    decache("../../../middleware/session-middleware");
    decache("../../../../test/helpers/session-helper");
    process.env.SUPPORT_MFA_RESET_WITH_IPV = "1";
    process.env.ROUTE_USERS_TO_NEW_IPV_JOURNEY = "1";
    const sessionMiddleware = await import(
      "../../../middleware/session-middleware"
    );
    const { getPermittedJourneyForPath } = await import(
      "../../../../test/helpers/session-helper"
    );

    sinon
      .stub(sessionMiddleware, "validateSessionMiddleware")
      .callsFake(function (
        req: Request,
        res: Response,
        next: NextFunction
      ): void {
        //Because we're using supertest to update the state after a first request, we only specifically
        //set this up for the first request
        if (req.path === firstRequestPath) {
          res.locals.sessionId = "tDy103saszhcxbQq0-mjdzU854";
          res.locals.strategicAppChannel = channel === CHANNEL.STRATEGIC_APP;
          res.locals.webChannel = channel == CHANNEL.WEB;

          req.session.user = {
            email: "test@test.com",
            journey: getPermittedJourneyForPath(currentNextPath),
            mfaMethodType: "AUTH_APP",
            isAccountRecoveryPermitted: isAccountRecoveryPermitted,
          };

          req.session.client = {
            redirectUri: "http://test-redirect.gov.uk/callback",
          };
        }
        next();
      });

    return (await import("../../../app")).createApp();
  }
});
