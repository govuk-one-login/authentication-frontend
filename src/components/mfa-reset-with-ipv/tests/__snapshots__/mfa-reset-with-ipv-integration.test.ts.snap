// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mfa reset with ipv Open One login in browser get should not render the page when coming from an arbitrary page 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "sign in",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Mfa reset with ipv should render the guidance page for when someone is trying to access via an app journey 1`] = `
Object {
  "contentId": undefined,
  "taxonomyLevel1": undefined,
  "taxonomyLevel2": undefined,
  "taxonomyLevel3": undefined,
  "taxonomyLevel4": undefined,
  "taxonomyLevel5": undefined,
}
`;
